src/lib.rs

```diff
@@ -104,6 +107,124 @@ impl Invoice {
     fn version_in_range(&self, requirement: &str) -> bool {
         version_compare(self.bindle.id.version(), requirement)
     }
+
+    fn cleartext(&self, by: String, role: SignatureRole) -> String {
+        let id = self.bindle.id.clone();
+        let mut buf = vec![
+            by.clone(),
+            id.name().to_owned(),
+            id.version_string(),
+            format!("{:?}", role).to_lowercase(),
+            '~'.to_string(),
+        ];
+
+        // Add bindles
+        self.parcel.as_ref().map(|list| {
+            list.iter().for_each(|p| {
+                buf.push(p.label.sha256.clone());
+            })
+        });
+
+        buf.join("\n")
+    }
+
+    /// Sign the parcels on the current package.
+    ///
+    /// Note that this signature will be invalidated if any parcels are
+    /// added after this signature.
+    ///
+    /// In the current version of the spec, a signature is generated by combining the
+    /// signer's ID, the invoice version, and a list of parcels, and then performing
+    /// a cryptographic signature on those fields. The result is then stored in
+    /// a `[[siganture]]` block on the invoice. Multiple signatures can be attached
+    /// to any invoice.
+    pub fn sign(
+        &mut self,
+        signer_name: String,
+        signer_role: SignatureRole,
+        key: Keypair,
+    ) -> Result<(), SignatureError> {
+        let cleartext = self.cleartext(signer_name.clone(), signer_role.clone());
+        let signature: EdSignature = key.sign(cleartext.as_bytes());
+
+        let signature_entry = Signature {
+            by: signer_name,
+            key: base64::encode(key.public.to_bytes()),
+            signature: base64::encode(signature.to_bytes()),
+            role: signer_role,
+        };
+
+        match self.signature.as_mut() {
+            Some(signatures) => signatures.push(signature_entry),
+            None => self.signature = Some(vec![signature_entry]),
+        };
+
+        Ok(())
+    }
+
+    /// Verify that every signature on this invoice is correct.
+    ///
+    /// The signature block is considered safe iff:
+    /// - All of the signatures are verified
+    /// - At least one of the signatures is made with a key on the keyring
+    ///
+    /// To verify a signature, the keyring is loaded, and then for each
+    /// `[[signature]]` object on an invoice, the ciphertext is verified.
+    /// The cleartext can be reconstructed from the invoice itself.
+    ///
+    /// Note that the purpose of the keyring is to ensure that we know about the
+    /// entity that claims to have signed the invoice.
+    ///
+    /// If _any_ signature fails, this should be considered a fatal error.
```

Comment:
I would also document in here that if no signature blocks are present, this will succeed

Reply:
Fixed